check_source_in_factory.py

```diff
@@ -67,28 +67,37 @@ def check_source_submission(self, src_project, src_package, src_rev, target_proj
             # handle here to avoid crashing on the next line
             self.logger.info("Could not get source info for %s/%s@%s" % (src_project, src_package, src_rev))
             return False
-        project = self._package_get_upstream_project(target_package)
-        if project is None:
-            self.logger.error("no upstream project found for {}, can't check".format(package))
+        projects = self._package_get_upstream_projects(target_package)
+        if projects is None:
+            self.logger.error("no upstream project found for {}, can't check".format(target_package))
             return False
 
-        good = self._check_project(project, target_package, src_srcinfo.verifymd5)
-        if good:
-            self.logger.info("{} is in {}".format(target_package, project))
-            return good
+        self.review_messages['declined'] = 'the package needs to be accepted in {} first'.format(' or '.join(projects))
+        for project in projects:
+            self.logger.info("Checking in project %s" % project)
+            good = self._check_project(project, target_package, src_srcinfo.verifymd5)
+            if good:
+                self.logger.info("{} is in {}".format(target_package, project))
+                return good
 
-        good = self._check_requests(project, target_package, src_srcinfo.verifymd5)
-        if good:
-            self.logger.info("{} already reviewed for {}".format(target_package, project))
+            good = self._check_requests(project, target_package, src_srcinfo.verifymd5)
+            if good:
+                self.logger.info("{} already reviewed for {}".format(target_package, project))
+
+        if not good:
+            self.logger.info('{} failed source submission check'.format(target_package))
 
         return good
 
-    def _package_get_upstream_project(self, package):
-        """ return project where the specified pacakge is supposed to come
+    def _package_get_upstream_projects(self, package):
+        """ return list of projects where the specified package is supposed to come
         from. Either by lookup table or self.factory """
         if self.lookup and package in self.lookup:
             return self.lookup[package]
 
+        if type(self.factory) is str:
+            return [self.factory]
```

Comment:
Actually this check is in the wrong place because the code just above it can return a string to the caller expecting a list. The value from the lookup needs to be converted also. I think it's best to keep this check and conversion in the _package_get_upstream_projects method and keep the change hidden from other users of the FactorySourceChecker class (like check_tags_in_requests.py).